src/fsw/FCCode/Drivers/Piksi.cpp

```diff
@@ -271,32 +307,34 @@ unsigned char Piksi::read_all() {
     #ifdef DESKTOP
     return _read_return;
     #else
+
+    noInterrupts();
+
+    int bytes = Serial4.available();
+    last_bytes = 0;
+
+    buffer_begin = buffer;
+    buffer_end = buffer;
+    // copy buffer up to available
+    for (int i = 0; i < bytes; i++){
+        *(buffer_end++) = Serial4.read();
+    }
+
+    microdelta = interrupt_count*100; //convert to us
+
+    interrupts();
+
     
     _gps_time_update = false;
     _pos_ecef_update = false;
     _vel_ecef_update = false;
     _baseline_ecef_update = false;
-
-    int initial_time = micros();
     
-    if(bytes_available()){
-        bool crc_error = false;
-        while(bytes_available() && (micros() - initial_time < READ_ALL_LIMIT)){
-            //call process_buffer() to process data, and check if crc_error happened
-            if(process_buffer() < 0)
-                crc_error = true;
-        }
-
-        //ensure that if the while loop terminated because of exceeding the READ_ALL_LIMIT
-        //it will enter the clear bytes condition below
-        delayMicroseconds(5);
-
-        if(micros()-initial_time >= READ_ALL_LIMIT){
-            clear_bytes();
-            return 5;
+    if(buffer_begin < buffer_end){ 
+        bool crc_error;
+        while(buffer_begin < buffer_end){
+            crc_error = process_buffer() < 0;
         }

```

Comment:
Should we make this or equals @shihaocao? Unfamiliar with how Piksi works but it seems safer...
```suggestion
        bool crc_error = false;
        while(buffer_begin < buffer_end){
            crc_error |= process_buffer() < 0;
        }

```

Reply:
I think `|=` is the safer option because if any particular message has a crc_error, its probably safer to toss the whole read out.

Reply:
src/fsw/FCCode/Drivers/Piksi.cpp:336:46: error: 'crc_error' may be used uninitialized in this function [-Werror=maybe-uninitialized]
             crc_error |= process_buffer() < 0;